plugins/woocommerce/client/blocks/assets/js/atomic/blocks/product-elements/image/block.tsx

```diff
@@ -101,6 +102,28 @@ const Image = ( {
 type Props = BlockAttributes &
 	HTMLAttributes< HTMLDivElement > & { style?: Record< string, unknown > };
 
+// props.product is not listed in the BlockAttributes explicitly,
+// but it is implicitly passed from the All Products block.
+// This is what distinguishes this block from the other usage of the Product Image component.
+const displayLegacySaleBadge = ( props: Props ) => {
```

Comment:
In this case I think we should still type `product`, maybe in a type called: `LegacyProps`. The props object should be aware of anything that _could_ be passed to it.

Reply:
Done https://github.com/woocommerce/woocommerce/pull/57889/commits/f151536b1e085fa6ab1dd9fccee86c7afcf8f21d

Reply:
Mmm… interesting way to type it, I find it a bit noisy, why did you intersect two different objects (one for `style` and one for `product`) instead of one object for both props?

Also, since you are using `Props` to type the Block below, `product` is expected to be defined in the normal block, which is not always the case. Either `product` needs to be optional, or I would separate the types into `Props` and `LegacyProps` for even further clarity, so future consumers don't think it's ok to send `product` to the badge even when “technically” accepted.

Reply:
> Mmm… interesting way to type it, I find it a bit noisy, why did you intersect two different objects (one for style and one for product) instead of one object for both props?

I have nothing to defend myself, honestly, I don't know, I think I rushed 😅 

> I would separate the types into Props and LegacyProps for even further clarity, so future consumers don't think it's ok to send product to the badge even when “technically” accepted.

Done in https://github.com/woocommerce/woocommerce/pull/57889/commits/41e988bc05b51d1c8e6e29bcf002fe34b66aeb74. That's what you meant, right? 🙌